### ISSUE TITLE
Using `to_parquet` with `pyarrow` fails on compression argument

### ISSUE THREAD

**Issue by jsnowacki** [dask/dask#2985]
I was saving DataFrame to Parquet using Arrow as the engine fails if there is `compression` argument passed to the function. Indeed, looking into [Arrow's documentation](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_metadata.html#pyarrow.parquet.write_metadata) there is no such method, nor `kwargs`, thus, the below fail. Interestingly, it fails in the middle of computations, not at the beginning. Note that I had only 'pyarrow' installed, thus, no engine has been specified explicitly. The below works after switching engine to `fastparquet`.

Using:
* python 3.6 via Anaconda
* dask 0.16.0
* pyarrow 0.7.1

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-55-c751a9021c02> in <module>()
      2 
      3 with ProgressBar():
----> 4     posts.repartition(npartitions=200)        .to_parquet('large_data/Posts.parq', compression='snappy')

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\core.py in to_parquet(self, path, *args, **kwargs)
    983         """ See dd.to_parquet docstring for more information """
    984         from .io import to_parquet
--> 985         return to_parquet(self, path, *args, **kwargs)
    986 
    987     def to_csv(self, filename, **kwargs):

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\io\parquet.py in to_parquet(df, path, engine, compression, write_index, append, ignore_divisions, partition_on, storage_options, compute, **kwargs)
    616 
    617     if compute:
--> 618         out.compute()
    619         return None
    620     return out

C:\ProgramData\Anaconda3\lib\site-packages\dask\base.py in compute(self, **kwargs)
    133         dask.base.compute
    134         """
--> 135         (result,) = compute(self, traverse=False, **kwargs)
    136         return result
    137 

C:\ProgramData\Anaconda3\lib\site-packages\dask\base.py in compute(*args, **kwargs)
    331     postcomputes = [a.__dask_postcompute__() if is_dask_collection(a)
    332                     else (None, a) for a in args]
--> 333     results = get(dsk, keys, **kwargs)
    334     results_iter = iter(results)
    335     return tuple(a if f is None else f(next(results_iter), *a)

C:\ProgramData\Anaconda3\lib\site-packages\dask\multiprocessing.py in get(dsk, keys, num_workers, func_loads, func_dumps, optimize_graph, **kwargs)
    175                            get_id=_process_get_id, dumps=dumps, loads=loads,
    176                            pack_exception=pack_exception,
--> 177                            raise_exception=reraise, **kwargs)
    178     finally:
    179         if cleanup:

C:\ProgramData\Anaconda3\lib\site-packages\dask\local.py in get_async(apply_async, num_workers, dsk, result, cache, get_id, rerun_exceptions_locally, pack_exception, raise_exception, callbacks, dumps, loads, **kwargs)
    519                         _execute_task(task, data)  # Re-execute locally
    520                     else:
--> 521                         raise_exception(exc, tb)
    522                 res, worker_id = loads(res_info)
    523                 state['cache'][key] = res

C:\ProgramData\Anaconda3\lib\site-packages\dask\compatibility.py in reraise(exc, tb)
     57     def reraise(exc, tb=None):
     58         if exc.__traceback__ is not tb:
---> 59             raise exc.with_traceback(tb)
     60         raise exc
     61 

C:\ProgramData\Anaconda3\lib\site-packages\dask\local.py in execute_task()
    288     try:
    289         task, data = loads(task_info)
--> 290         result = _execute_task(task, data)
    291         id = get_id()
    292         result = dumps((result, id))

C:\ProgramData\Anaconda3\lib\site-packages\dask\local.py in _execute_task()
    269         func, args = arg[0], arg[1:]
    270         args2 = [_execute_task(a, cache) for a in args]
--> 271         return func(*args2)
    272     elif not ishashable(arg):
    273         return arg

C:\ProgramData\Anaconda3\lib\site-packages\dask\compatibility.py in apply()
     45     def apply(func, args, kwargs=None):
     46         if kwargs:
---> 47             return func(*args, **kwargs)
     48         else:
     49             return func(*args)

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\io\parquet.py in _write_partition_pyarrow()
    417     if metadata_path is not None:
    418         with open_with(metadata_path, 'wb') as fil:
--> 419             parquet.write_metadata(t.schema, fil, **kwargs)
    420 
    421 

TypeError: write_metadata() got an unexpected keyword argument 'compression'
```

**Comment by xhochy** [comment]
I think this was fixed in `dask` master. Can you try if installing `dask` from git solves the issue?

**Comment by jsnowacki** [comment]
I've tested it and it's fixed, i.e. the file writing goes through. Though, I get a different behavior. First there is no `_metadata` file saved, only `_common_metadata`; before it was attempting to save it, though, it failed. Second I get the following errors when I try to read the files using engine `fastparquet`; reading works fine with `pyarrow`. I don't know if this permission error is due to `_metadata` missing or something else, but nothing else is using the files.

```---------------------------------------------------------------------------
FileNotFoundError                         Traceback (most recent call last)
C:\ProgramData\Anaconda3\lib\site-packages\fastparquet\api.py in __init__(self, fn, verify, open_with, sep, root)
     95                 self.fn = fn2
---> 96                 with open_with(fn2, 'rb') as f:
     97                     self._parse_header(f, verify)

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\core.py in __enter__(self)
    311         mode = self.mode.replace('t', '').replace('b', '') + 'b'
--> 312         f = f2 = self.myopen(self.path, mode=mode)
    313         CompressFile = merge(seekable_files, compress_files)[self.compression]

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\local.py in open(self, path, mode, **kwargs)
     60         path = self._trim_filename(path)
---> 61         return open(path, mode=mode)
     62 

FileNotFoundError: [Errno 2] No such file or directory: 'D:/Kuba/Projects/large_data/Posts.parq\\_metadata\\_metadata'

During handling of the above exception, another exception occurred:

FileNotFoundError                         Traceback (most recent call last)
C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\io\parquet.py in _read_fastparquet(fs, paths, myopen, columns, filters, categories, index, storage_options)
     57                                          open_with=myopen,
---> 58                                          sep=fs.sep)
     59         except Exception:

C:\ProgramData\Anaconda3\lib\site-packages\fastparquet\api.py in __init__(self, fn, verify, open_with, sep, root)
    100                 self.fn = fn
--> 101                 with open_with(fn, 'rb') as f:
    102                     self._parse_header(f, verify)

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\core.py in __enter__(self)
    311         mode = self.mode.replace('t', '').replace('b', '') + 'b'
--> 312         f = f2 = self.myopen(self.path, mode=mode)
    313         CompressFile = merge(seekable_files, compress_files)[self.compression]

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\local.py in open(self, path, mode, **kwargs)
     60         path = self._trim_filename(path)
---> 61         return open(path, mode=mode)
     62 

FileNotFoundError: [Errno 2] No such file or directory: 'D:/Kuba/Projects/large_data/Posts.parq\\_metadata'

During handling of the above exception, another exception occurred:

FileNotFoundError                         Traceback (most recent call last)
C:\ProgramData\Anaconda3\lib\site-packages\fastparquet\api.py in __init__(self, fn, verify, open_with, sep, root)
     95                 self.fn = fn2
---> 96                 with open_with(fn2, 'rb') as f:
     97                     self._parse_header(f, verify)

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\core.py in __enter__(self)
    311         mode = self.mode.replace('t', '').replace('b', '') + 'b'
--> 312         f = f2 = self.myopen(self.path, mode=mode)
    313         CompressFile = merge(seekable_files, compress_files)[self.compression]

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\local.py in open(self, path, mode, **kwargs)
     60         path = self._trim_filename(path)
---> 61         return open(path, mode=mode)
     62 

FileNotFoundError: [Errno 2] No such file or directory: 'D:/Kuba/Projects/large_data/Posts.parq\\_metadata'

During handling of the above exception, another exception occurred:

PermissionError                           Traceback (most recent call last)
<ipython-input-2-c4cf5982d1c9> in <module>()
----> 1 posts = dd.read_parquet('D:/Kuba/Projects/large_data/Posts.parq', engine='fastparquet')
      2 posts

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\io\parquet.py in read_parquet(path, columns, filters, categories, index, storage_options, engine)
    536 
    537     return read(fs, paths, file_opener, columns=columns, filters=filters,
--> 538                 categories=categories, index=index)
    539 
    540 

C:\ProgramData\Anaconda3\lib\site-packages\dask\dataframe\io\parquet.py in _read_fastparquet(fs, paths, myopen, columns, filters, categories, index, storage_options)
     58                                          sep=fs.sep)
     59         except Exception:
---> 60             pf = fastparquet.ParquetFile(paths[0], open_with=myopen, sep=fs.sep)
     61 
     62     check_column_names(pf.columns, categories)

C:\ProgramData\Anaconda3\lib\site-packages\fastparquet\api.py in __init__(self, fn, verify, open_with, sep, root)
     99             except (IOError, OSError):
    100                 self.fn = fn
--> 101                 with open_with(fn, 'rb') as f:
    102                     self._parse_header(f, verify)
    103         self.open = open_with

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\core.py in __enter__(self)
    310     def __enter__(self):
    311         mode = self.mode.replace('t', '').replace('b', '') + 'b'
--> 312         f = f2 = self.myopen(self.path, mode=mode)
    313         CompressFile = merge(seekable_files, compress_files)[self.compression]
    314         if PY2:

C:\ProgramData\Anaconda3\lib\site-packages\dask\bytes\local.py in open(self, path, mode, **kwargs)
     59         """
     60         path = self._trim_filename(path)
---> 61         return open(path, mode=mode)
     62 
     63     def ukey(self, path):

PermissionError: [Errno 13] Permission denied: 'D:/Kuba/Projects/large_data/Posts.parq'
```
Moreover, the below fix removes only the `compression` parameter.
https://github.com/dask/dask/blob/9b7e21cfc33cb9c66abb85b547669edb915bc9e7/dask/dataframe/io/parquet.py#L416.
When some other parameter would sneak into `kwargs`, you may get a similar error as function `write_metadata` takes only certain named arguments and no `kwargs`. Thus, this should only take out the arguments from `kwargs` that are handeled by the function. I can propose a PR to fix that.

**Comment by jcrist** [comment]
> First there is no _metadata file saved, only _common_metadata; before it was attempting to save it, though, it failed.

Pyarrow currently doesn't have the machinery to write a `_metadata` file, only `_common_metadata`. Before we were erroneously writing `_common_metadata` as `_metadata`.

> Second I get the following errors when I try to read the files using engine fastparquet; reading works fine with pyarrow.

`fastparquet` requires `_metadata` if you point only at a directory (and thus can't read pyarrow written parquet as easily). You can use a glob to read with `fastparquet` and things should work fine:


```python
dd.read_parquet('mypath/*.parq')
```

> Moreover, the below fix removes only the compression parameter.

That's a good point, PR welcome.

**Comment by jcrist** [comment]
Fixed by #2986. Closing.
